### PR TITLE
fix: resolve target @Joined fields through relation lookups

### LINKED ISSUE #14: Query.Foreign does not resolve @Joined fields on the relation target (label comes back empty in list)
## Summary

When a relation (`Foreign`) plucks a field from its target, and that field is itself a `@Joined` field (materialized from a second table), `Query.Foreign` returns it empty/missing. The relation value is therefore unusable downstream (a UI rendering the relation shows an empty cell).

The **same field resolves correctly** when the target is fetched through its own `list`/`select` route, because `Query.Joined` is applied there. The bug is specific to resolving the field **as the target of a relation**.

## Affected version

`@antelopejs/interface-data-api@0.1.4` (default branch `main`).

## Root cause

`Query.Foreign` resolves a relation against the **raw** target table and plucks the requested fields — `src/components.ts:249-314`:

```ts
const other = resolveSchemaDb(db, schemaName).table(table);   // raw target table
if (pluckField) {
  query = query.lookup(
    other.pluck("_internal", ...pluckField) as Table<any>,
    name,
    index || "_id",
  );
}
```

`other` is the raw table. The target's `@Joined` fields are materialized by a **separate** function, `Query.Joined` (`src/components.ts:356+`), which is only ever applied to the target's *own* query — never to this sub-lookup. So any plucked field that is a `@Joined` (a non-physical column) comes back empty, and the relation value is broken.

The singular (`Datum`) branch has the same gap — `src/components.ts:303-307`:

```ts
changedFields[name] = Get(remoteDb.table(table), obj.key(name), index).default(null);
// pluck on the raw table; target @Joined fields are never resolved here
```

## Reproduction (decorators)

Target controller whose **label field is `@Joined`**:

```ts
class MemberAPI extends DataController(Member, ...) {
  @Select() @Listable() declare _id: string;

  @Listable() declare userId: string;                  // alt key, NOT @Select

  @Select(["userId"]) @Listable(["userId"])
  @Joined({ table: User, localKey: "userId", remoteField: "name" })
  declare name: string;                                // ← label is a @Joined column
}
```

A mapping controller referencing it through a relation that plucks `name`
(i.e. `Foreign(MemberTable, "userId", false, ["_id", "name"])`):

```ts
@Foreign(MemberTable, "userId", false, ["_id", "name"])
declare member: string;
```

Observed:

- `GET /member/list` → `{ "_id": "m-1", "userId": "u-alice", "name": "Alice Martin" }` ✅ (`Query.Joined` applied)
- `GET /mapping/list` → the `member` relation column comes back **empty / missing** ❌ (`Query.Foreign`, no join applied)

A **control** relation whose target label is a **native** column resolves correctly — including when joined on a non-`_id` `index` (e.g. `Foreign(DeptTable, "code", false, ["_id", "name"])` where `name` is a real column). So the trigger is precisely: **a plucked/label field on the relation target is a `@Joined` field.** The `index`/non-`_id` key is *not* the cause.

## Expected

`Query.Foreign` should materialize the target's `@Joined` fields before/while plucking, so a relation can use a joined field as its label — consistent with what the target's own `list`/`select` already returns.

## Suggested fix

Apply the target's joined groups inside the `Query.Foreign` sub-lookup (both the `Stream` and `Datum` branches). The target class is available in `field.foreign` (`_tableClass`), so its `DataAPIMeta` — and therefore its joined groups (`collectJoinedGroups`) — can be retrieved and applied to `other` before `.pluck(...)`.

A regression test could extend `src/tests/components/joined.test.ts` with a relation that plucks a joined label and asserts the label is present in the parent's `list` output.

### PR DESCRIPTION
**Pull request by MrSociety404**
## What

Fixes #14 — a relation (`Foreign`) whose label (`keyMapping`) points at a `@Joined`
field on the target controller came back **empty** in `list`/`get`, while the
same field resolves fine when the target is fetched through its own route.

## Root cause

`Query.Foreign` resolved a relation against the **raw** target table and plucked
the requested fields:

```ts
const other = resolveSchemaDb(db, schemaName).table(table);
query = query.lookup(other.pluck("_internal", ...pluckField) as Table<any>, name, index || "_id");
```

A `@Joined` field is not a physical column — it is materialized by `Query.Joined`,
which is only ever applied to the target's *own* query, never to this sub-lookup.
So a joined label resolved to nothing. The `Datum` (get) branch had the same gap.

## Fix

- **`metadata.ts`** — the `Foreign` decorator / `setForeign` now accept an optional
  `targetMeta` (the target's `DataAPIMeta`). When given, the target's relevant
  `@Joined` fields (those present in `pluck`) are captured into `field.foreign`
  as `ForeignJoinedRef[]`, and their `localKey` is added to the pluck so the join
  key is fetched. Fully backward compatible: without `targetMeta`, behaviour is
  unchanged.
- **`components.ts`** — `Query.Foreign` re-materializes those joined fields after
  the lookup, in both the `Stream` (list) and `Datum` (get) paths, mirroring the
  hoist→lookup→merge approach of `Query.Joined`. Orphan foreign keys still resolve
  to `null`.

This is the framework half of a two-part fix: a downstream consumer that builds
relations via `Foreign` (e.g. a `RelationType` whose target exposes a joined
label) passes the target meta to opt in. The non-`_id` join `index` was **not**
the cause and keeps working.

## Tests

New `src/tests/components/foreign_joined.test.ts` (target whose `name` label is a
`@Joined` from another table, referenced through a relation):

- resolves the joined label through a relation in **list** ✅ (the reported bug)
- resolves the joined label through a relation in **get** ✅
- returns `null` for an **orphan** foreign key ✅

Full suite green locally (`72 passing`), lint and build clean.

<h3>Greptile Summary</h3>

This PR fixes a bug where a `@Foreign` relation whose pluck included a `@Joined` field on the target controller would return that field empty, because the plain table lookup never materialized virtual/joined columns. The fix threads the target's `DataAPIMeta` into `setForeign`, captures the relevant `@Joined` descriptors as `ForeignJoinedRef[]`, and re-materializes them at query time via a second lookup in both the stream (list) and datum (get) paths.

- **`metadata.ts`**: `setForeign` now accepts an optional `targetMeta`; joined fields present in the requested pluck are captured into `foreign[6]`, with their `localKey` kept out of `foreign[4]` to prevent leaking join keys into the API response.
- **`components.ts`**: `Query.Foreign` extends the lookup pluck with join keys at query time, calls new `applyForeignJoinedStream`/`applyForeignJoinedDatum` helpers, and `Validation.ClearInternal` gains orphan-detection logic to normalize a joined-but-unmatched object back to `null`.
- **`foreign_joined.test.ts`**: New test suite covering list, get, and orphan scenarios in both paths.

<h3>Confidence Score: 5/5</h3>

Safe to merge. The change is strictly additive: without targetMeta the behaviour is identical to before, and new code paths are exercised only when opt-in metadata is supplied.

The happy-path logic for both stream and datum paths is sound and well-tested across list, get, and orphan scenarios. The two findings are narrow edge cases that do not affect the bug being fixed or any currently tested configuration.

The orphan-normalization branch in src/components.ts (ClearInternal, lines 919-931) is worth a second look for the all-joined-pluck edge case. The ShelfListed interface in the test file should be corrected to match the actual response shape.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/metadata.ts | Adds ForeignJoinedRef interface and extends setForeign/Foreign decorator with optional targetMeta to capture target @Joined fields at decoration time. Logic is correct: joined fields present in the pluck are collected into targetJoined, their localKey is intentionally excluded from foreign[4] to avoid leaking join keys in the API response. |
| src/components.ts | Adds applyForeignJoinedStream and applyForeignJoinedDatum helpers and wires them into Query.Foreign; extends Validation.ClearInternal with orphan detection logic. Stream and datum paths for the happy path look correct; minor gap in orphan normalization when all pluck fields are joined fields. |
| src/tests/components/foreign_joined.test.ts | New test covering list, get, and orphan cases for the fixed bug. Assertions are thorough; minor issue: the ShelfListed TypeScript interface includes authorId which the test itself asserts is absent from the response. |

</details>

</details>

<h3>Sequence Diagram</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Client
    participant QF as Query.Foreign
    participant Lookup1 as DB lookup book table
    participant FJS as applyForeignJoinedStream
    participant Lookup2 as DB lookup author table
    participant CI as Validation.ClearInternal

    Client->>QF: list shelves
    QF->>Lookup1: lookup book pluck _id title authorId
    Lookup1-->>QF: stream with book _id title authorId
    QF->>FJS: applyForeignJoinedStream for authorName
    FJS->>FJS: map hoist book.authorId into tmpKey
    FJS->>Lookup2: lookup author table by tmpKey
    Lookup2-->>FJS: author row resolved
    FJS->>FJS: map merge authorName then remove tmpKey
    FJS-->>QF: stream with book _id title authorId authorName
    QF-->>CI: raw results
    CI->>CI: orphan check physicalKeys all null returns null
    CI->>CI: pluck to _id title authorName strips authorId
    CI-->>Client: results with book _id title authorName
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Client
    participant QF as Query.Foreign
    participant Lookup1 as DB lookup book table
    participant FJS as applyForeignJoinedStream
    participant Lookup2 as DB lookup author table
    participant CI as Validation.ClearInternal

    Client->>QF: list shelves
    QF->>Lookup1: lookup book pluck _id title authorId
    Lookup1-->>QF: stream with book _id title authorId
    QF->>FJS: applyForeignJoinedStream for authorName
    FJS->>FJS: map hoist book.authorId into tmpKey
    FJS->>Lookup2: lookup author table by tmpKey
    Lookup2-->>FJS: author row resolved
    FJS->>FJS: map merge authorName then remove tmpKey
    FJS-->>QF: stream with book _id title authorId authorName
    QF-->>CI: raw results
    CI->>CI: orphan check physicalKeys all null returns null
    CI->>CI: pluck to _id title authorName strips authorId
    CI-->>Client: results with book _id title authorName
```

</a>

<sub>Reviews (2): Last reviewed commit: ["fix: address review — no join-key leak, ..."](https://github.com/antelopejs/interface-data-api/commit/a1c6f54718d736942ecf60a053333074c46fa974) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=39013753)</sub>

### CODE REVIEW

File: src/components.ts [hunk 1]
```diff
@@ -313,6 +343,83 @@ export namespace Query {
...
+            row.key(group.localKey) as ValueProxy<string>,
+            group.remoteIndex,
+          )
+            .key(group.remoteField)
+            .default(null),
+        }),
+      ) as Datum<any>;
+    }
+    return d;
+  }
```
[the remark this answers is not in the document]
Comment (MrSociety404) [reply src/components.ts:434]: Good catch on the missing coverage — added a `get` + orphan test (`returnsNullForOrphanRelationInGet`). On the Mongo backend the `.do()` over a null datum short-circuits to null (no `null.merge` evaluation), so it resolves to `null` rather than throwing. Verified green. Fixed in a1c6f54.

File: src/components.ts [hunk 2]
```diff
@@ -313,6 +343,83 @@ export namespace Query {
...
+                [group.name]: row
+                  .key(tmpKey)
+                  .default({})
+                  .key(group.remoteField)
+                  .default(null),
+              }),
+          }),
+        )
+        .without(tmpKey) as Stream<any>;
+    }
```
[the remark this answers is not in the document]
Comment (MrSociety404) [reply src/components.ts:401]: Confirmed and fixed. `branch`/value-level conditionals aren't available in this DSL (`and`/`or` return booleans), so the null can't be preserved in-query for the stream path. Instead `ClearInternal` now normalizes a joined-orphan (object with only null @Joined fields and no physical pluck fields) back to a null relation — consistent with the get path. Test `returnsNullForOrphanRelation` now asserts strict `book === null`. Fixed in a1c6f54.

File: src/metadata.ts [hunk 3]
```diff
@@ -583,20 +598,58 @@ export class DataAPIMeta {
...
+        });
+        extraKeys.add(j.localKey);
+      }
+      if (groups.length > 0) {
+        targetJoined = groups;
+        if (pluck) {
+          finalPluck = Array.from(new Set([...pluck, ...extraKeys]));
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (MrSociety404) [reply src/metadata.ts:640]: Fixed. `setForeign` no longer appends the join `localKey` to the stored pluck (`foreign[4]`). `Query.Foreign` extends only the *lookup* pluck with the join keys at query time, so they're fetched for the second lookup but stripped from the response by `ClearInternal`. Test now asserts `Object.keys(book)` is exactly `[_id, authorName, title]` (no `authorId`). Fixed in a1c6f54.

**Comment by Thomasims** [comment]
@greptile review

**Review by Thomasims — APPROVED** [review]
